File: app/src/main/java/com/demo/railbridge/bridge/BridgeResponseFactory.java
package com.demo.railbridge.bridge;

import com.demo.railbridge.sdk.BalanceSnapshot;
import com.demo.railbridge.sdk.ChargeResult;
import com.demo.railbridge.sdk.SdkErrorCode;
import com.demo.railbridge.sdk.SdkStatusSnapshot;

import org.json.JSONException;
import org.json.JSONObject;

public final class BridgeResponseFactory {

    private BridgeResponseFactory() {
    }

    public static String buildChargeSuccess(ChargeResult result, Metadata metadata) {
        try {
            JSONObject data = new JSONObject();
            data.put("transactionId", result.getTransactionId());
            data.put("amount", result.getAmount());
            data.put("balance", result.getBalance());
            data.put("timestamp", result.getTimestamp());
            return buildSuccess("requestCharge", data, 0, metadata);
        } catch (JSONException e) {
            return buildError("requestCharge", SdkErrorCode.ERR_UNKNOWN, 0, null, null, null, metadata);
        }
    }

    public static String buildBalanceSuccess(BalanceSnapshot result, Metadata metadata) {
        try {
            JSONObject data = new JSONObject();
            data.put("cardId", result.getCardId());
            data.put("balance", result.getBalance());
            data.put("timestamp", result.getTimestamp());
            return buildSuccess("getBalance", data, 0, metadata);
        } catch (JSONException e) {
            return buildError("getBalance", SdkErrorCode.ERR_UNKNOWN, 0, null, null, null, metadata);
        }
    }

    public static String buildStatusSuccess(SdkStatusSnapshot result, Metadata metadata) {
        try {
            JSONObject data = new JSONObject();
            data.put("initialized", result.isInitialized());
            data.put("version", result.getVersion());
            return buildSuccess("getSdkStatus", data, 0, metadata);
        } catch (JSONException e) {
            return buildError("getSdkStatus", SdkErrorCode.ERR_UNKNOWN, 0, null, null, null, metadata);
        }
    }

    public static String buildReportAck(String code, String message, Metadata metadata) {
        try {
            JSONObject response = new JSONObject();
            response.put("status", "received");
            response.put("message", message);
            response.put("code", code);
            addMetadata(response, metadata);
            return response.toString();
        } catch (JSONException e) {
            return "{\"status\":\"received\",\"message\":\"Error report received\"}";
        }
    }

    public static String attachMetadata(String resultJson, Metadata metadata) {
        if (resultJson == null || resultJson.trim().isEmpty() || metadata == null) {
            return resultJson;
        }

        try {
            JSONObject response = new JSONObject(resultJson);
            addMetadata(response, metadata);
            return response.toString();
        } catch (JSONException e) {
            return resultJson;
        }
    }

    public static String buildError(
            String method,
            SdkErrorCode errorCode,
            int retryCount,
            String logId,
            String vendorCode,
            Boolean retryable,
            Metadata metadata
    ) {
        try {
            JSONObject error = new JSONObject();
            error.put("code", errorCode.getCode());
            error.put("message", errorCode.getMessage());
            error.put("retryCount", retryCount);
            error.put("resolved", false);

            if (vendorCode != null && !vendorCode.trim().isEmpty()) {
                error.put("vendorCode", vendorCode);
            }
            if (retryable != null) {
                error.put("retryable", retryable);
            }

            JSONObject response = new JSONObject();
            response.put("status", "error");
            response.put("method", method);
            response.put("error", error);
            if (logId != null && !logId.trim().isEmpty()) {
                response.put("logId", logId);
            }
            addMetadata(response, metadata);
            return response.toString();
        } catch (JSONException e) {
            return "{\"status\":\"error\",\"method\":\"" + method + "\"}";
        }
    }

    private static String buildSuccess(String method, JSONObject data, int retryCount, Metadata metadata)
            throws JSONException {
        JSONObject response = new JSONObject();
        response.put("status", "success");
        response.put("method", method);
        response.put("data", data);
        response.put("retryCount", retryCount);
        addMetadata(response, metadata);
        return response.toString();
    }

    private static void addMetadata(JSONObject response, Metadata metadata) throws JSONException {
        if (metadata == null) {
            return;
        }
        if (hasText(metadata.getCallbackId())) {
            response.put("callbackId", metadata.getCallbackId());
        }
        if (hasText(metadata.getCorrelationId())) {
            response.put("correlationId", metadata.getCorrelationId());
        }
        if (hasText(metadata.getPlatform())) {
            response.put("platform", metadata.getPlatform());
        }
        if (hasText(metadata.getStage())) {
            response.put("stage", metadata.getStage());
        }
        if (metadata.getDurationMs() != null) {
            response.put("durationMs", metadata.getDurationMs());
        }
        if (hasText(metadata.getScenario())) {
            response.put("scenario", metadata.getScenario());
        }
        if (hasText(metadata.getVendorCode())) {
            response.put("vendorCode", metadata.getVendorCode());
        }
        if (metadata.getRetryable() != null) {
            response.put("retryable", metadata.getRetryable());
        }
        if (metadata.getResolvedByRetry() != null) {
            response.put("resolvedByRetry", metadata.getResolvedByRetry());
        }
    }

    private static boolean hasText(String value) {
        return value != null && !value.trim().isEmpty();
    }

    public static final class Metadata {

        private final String callbackId;
        private final String correlationId;
        private final String platform;
        private final String stage;
        private final Long durationMs;
        private final String scenario;
        private final String vendorCode;
        private final Boolean retryable;
        private final Boolean resolvedByRetry;

        public Metadata(
                String callbackId,
                String correlationId,
                String platform,
                String stage,
                Long durationMs
        ) {
            this(callbackId, correlationId, platform, stage, durationMs, null, null, null, null);
        }

        public Metadata(
                String callbackId,
                String correlationId,
                String platform,
                String stage,
                Long durationMs,
                String scenario,
                String vendorCode,
                Boolean retryable,
                Boolean resolvedByRetry
        ) {
            this.callbackId = callbackId;
            this.correlationId = correlationId;
            this.platform = platform;
            this.stage = stage;
            this.durationMs = durationMs;
            this.scenario = scenario;
            this.vendorCode = vendorCode;
            this.retryable = retryable;
            this.resolvedByRetry = resolvedByRetry;
        }

        public String getCallbackId() {
            return callbackId;
        }

        public String getCorrelationId() {
            return correlationId;
        }

        public String getPlatform() {
            return platform;
        }

        public String getStage() {
            return stage;
        }

        public Long getDurationMs() {
            return durationMs;
        }

        public String getScenario() {
            return scenario;
        }

        public String getVendorCode() {
            return vendorCode;
        }

        public Boolean getRetryable() {
            return retryable;
        }

        public Boolean getResolvedByRetry() {
            return resolvedByRetry;
        }
    }
}

File: ios/RailBridgeIOS/RailBridgeIOS/Bridge/BridgeResponseFactory.swift
import Foundation

enum BridgeResponseFactory {
    struct Metadata: Equatable {
        let callbackId: String?
        let correlationId: String?
        let platform: String?
        let stage: String?
        let durationMs: Int?
        let scenario: String?
        let vendorCode: String?
        let retryable: Bool?
        let resolvedByRetry: Bool?
    }

    static func buildChargeSuccess(
        _ result: ChargeResult,
        retryCount: Int = 0,
        metadata: Metadata?
    ) -> String {
        let data: [String: Any] = [
            "transactionId": result.transactionId,
            "amount": result.amount,
            "balance": result.balance,
            "timestamp": result.timestamp
        ]
        return buildSuccess(method: "requestCharge", data: data, retryCount: retryCount, metadata: metadata)
    }

    static func buildBalanceSuccess(
        _ result: BalanceSnapshot,
        retryCount: Int = 0,
        metadata: Metadata?
    ) -> String {
        let data: [String: Any] = [
            "cardId": result.cardId,
            "balance": result.balance,
            "timestamp": result.timestamp
        ]
        return buildSuccess(method: "getBalance", data: data, retryCount: retryCount, metadata: metadata)
    }

    static func buildStatusSuccess(_ result: SdkStatusSnapshot, metadata: Metadata?) -> String {
        let data: [String: Any] = [
            "initialized": result.initialized,
            "version": result.version
        ]
        return buildSuccess(method: "getSdkStatus", data: data, retryCount: 0, metadata: metadata)
    }

    static func buildReportAck(code: String, message: String, metadata: Metadata?) -> String {
        var response: [String: Any] = [
            "status": "received",
            "message": message,
            "code": code
        ]
        mergeMetadata(into: &response, metadata: metadata)
        return stringify(response)
    }

    static func buildError(
        method: String,
        errorCode: SdkErrorCode,
        retryCount: Int,
        logId: String?,
        vendorCode: String?,
        retryable: Bool?,
        metadata: Metadata?
    ) -> String {
        var error: [String: Any] = [
            "code": errorCode.code,
            "message": errorCode.message,
            "retryCount": retryCount,
            "resolved": false
        ]
        if let vendorCode, !vendorCode.isEmpty {
            error["vendorCode"] = vendorCode
        }
        if let retryable {
            error["retryable"] = retryable
        }

        var response: [String: Any] = [
            "status": "error",
            "method": method,
            "error": error,
            "retryCount": retryCount
        ]
        if let logId, !logId.isEmpty {
            response["logId"] = logId
        }
        mergeMetadata(into: &response, metadata: metadata)
        return stringify(response)
    }

    private static func buildSuccess(
        method: String,
        data: [String: Any],
        retryCount: Int,
        metadata: Metadata?
    ) -> String {
        var response: [String: Any] = [
            "status": "success",
            "method": method,
            "data": data,
            "retryCount": retryCount
        ]
        mergeMetadata(into: &response, metadata: metadata)
        return stringify(response)
    }

    private static func mergeMetadata(into response: inout [String: Any], metadata: Metadata?) {
        guard let metadata else {
            return
        }

        response["callbackId"] = metadata.callbackId
        response["correlationId"] = metadata.correlationId
        response["platform"] = metadata.platform
        response["stage"] = metadata.stage
        response["durationMs"] = metadata.durationMs
        response["scenario"] = metadata.scenario
        response["vendorCode"] = metadata.vendorCode
        response["retryable"] = metadata.retryable
        response["resolvedByRetry"] = metadata.resolvedByRetry
        response = response.filter { _, value in !(value is NSNull) && !isNilOptional(value) }
    }

    private static func isNilOptional(_ value: Any) -> Bool {
        let mirror = Mirror(reflecting: value)
        return mirror.displayStyle == .optional && mirror.children.isEmpty
    }

    private static func stringify(_ dictionary: [String: Any]) -> String {
        guard
            JSONSerialization.isValidJSONObject(dictionary),
            let data = try? JSONSerialization.data(withJSONObject: dictionary, options: []),
            let string = String(data: data, encoding: .utf8)
        else {
            return "{}"
        }

        return string
    }
}
